src/main/java/com/google/research/bleth/simulator/AbstractSimulation.java

```diff
@@ -0,0 +1,278 @@
+    private final double radius;
+    private HashMap<String, Double> stats = new HashMap<>();
+
+    /** Returns the simulation's real board. */
+    RealBoard getBoard() {
+        return board;
+    }
+
+    /** Run entire simulation logic, including writing data to db. */
+    public void run() {
+        if (currentRound == 0) {
+            writeRoundState(); // round 0 is the initial simulation state
+            currentRound++;
+        }
+        while (currentRound <= maxNumberOfRounds) {
+            moveAgents();
+            updateObserversAwaknessState();
+            beaconsToObservers();
+            observersToResolver();
+            resolverEstimate();
+            writeRoundState();
+            updateSimulationStats();
+            currentRound++;
+        }
+        writeSimulationStats();
+    }
+
+    /** Move all agents according to their movement strategies and update the real board. */
+    void moveAgents() {
+        for (Beacon beacon : beacons) {
+            beacon.move();
+        }
+        for (Observer observer : observers) {
+            observer.move();
+        }
+    }
+
+    /** Update all observers awakeness states according to their awakeness strategies. */
+    void updateObserversAwaknessState() {
+        for (Observer observer : observers) {
+            observer.updateAwakenessState(currentRound);
+        }
+    }
+
+    /**
+     * Pass transmissions from beacons to observers while taking into consideration world-physics parameters,
+     * such as probability of transmission and distance between beacons and observers.
+     */
+    void beaconsToObservers() {
+        for (Beacon beacon : beacons) {
+            Transmission transmission = beacon.transmit();
+            for (Observer observer : observers) {
+                if (observer.isAwake()) {
+                    double distance = distance(beacon.getLocation(), observer.getLocation());
+                    if (distance <= radius) {
+                        observer.observe(transmission);
+                    }
+                }
+            }
+        }
+    }
+
+    /** Pass current-round information of transmission data from all observers to the simulation's resolver. */
+    void observersToResolver() {
+        for (Observer observer : observers) {
+            observer.passInformationToResolver();
+        }
+    }
+
+    /** Update resolver's estimated board. */
+    void resolverEstimate() { }
+
+    /** Write current-round state of the simulation to db. */
+    void writeRoundState() { }
+
+    /** Gather statistical data of the current round and update the aggregated simulation statistics based on all rounds. */
+    abstract void updateSimulationStats();
+
+    /** Write final simulation statistical data to db. */
+    void writeSimulationStats() { }
+
+    /** An abstract builder class designed to separate the construction of a simulation from its representation. */
+    public static abstract class Builder {
```

Comment:
Both _FreeBuilder_ and _AutoValue_ generate a builder with setters matching to the class attributes, i.e if the `AbstractSimulation` class has a `board` attribute, the generated builder will have a `setBoard()` method.
In our case, we want the builder to have `setRowNum()` and `setColNum()` methods, and during the `build()` it will create the `board` and `resolver` to be passed to the `AbstractSimulation` constructor.

(Same for other attributes, such as `beacons` and `observers` - we want the builder to have setter for the number of agents to create and their strategies, and the builder will create the agents itself).

Is there a way of using FreeBuilder or AutoValue to fit these requirements?

@OrpazFishel